ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/materialization/ks/SessionStoreCacheBypass.java

```diff
@@ -0,0 +1,199 @@
+    } catch (final NoSuchFieldException e) {
+      throw new RuntimeException("Stream internals changed unexpectedly!", e);
+    }
+  }
+
+  private SessionStoreCacheBypass() {}
+
+  interface SessionStoreCacheBypassFetcher {
+
+    KeyValueIterator<Windowed<Struct>, GenericRow> fetch(
+        ReadOnlySessionStore<Struct, GenericRow> store,
+        Struct key
+    );
+  }
+
+  @SuppressWarnings("unchecked")
+  public static KeyValueIterator<Windowed<Struct>, GenericRow> fetch(
+      final ReadOnlySessionStore<Struct, GenericRow> store,
+      final Struct key
+  ) {
+    Objects.requireNonNull(key, "key can't be null");
+
+    final StateStoreProvider provider;
+    final String storeName;
+    final QueryableStoreType<ReadOnlySessionStore<Struct, GenericRow>> storeType;
+    try {
+      provider = (StateStoreProvider) PROVIDER_FIELD.get(store);
+      storeName = (String) STORE_NAME_FIELD.get(store);
+      storeType = (QueryableStoreType<ReadOnlySessionStore<Struct, GenericRow>>)
+          STORE_TYPE_FIELD.get(store);
+    } catch (final IllegalAccessException e) {
+      throw new RuntimeException("Stream internals changed unexpectedly!", e);
+    }
+    final List<ReadOnlySessionStore<Struct, GenericRow>> stores
+        = provider.stores(storeName, storeType);
+    for (final ReadOnlySessionStore<Struct, GenericRow> sessionStore : stores) {
+      try {
+        final KeyValueIterator<Windowed<Struct>, GenericRow> result
+            = fetchUncached(sessionStore, key);
+        // returns the first non-empty iterator
+        if (!result.hasNext()) {
+          result.close();
+        } else {
+          return result;
+        }
+      } catch (final InvalidStateStoreException e) {
+        throw new InvalidStateStoreException(
+            "State store is not available anymore and may have been migrated to another instance; "
+                + "please re-discover its location from the state metadata.", e);
+      }
+    }
+    return new EmptyKeyValueIterator();
+  }
+
+  @SuppressWarnings("unchecked")
+  private static KeyValueIterator<Windowed<Struct>, GenericRow> fetchUncached(
+      final ReadOnlySessionStore<Struct, GenericRow> sessionStore,
+      final Struct key
+  ) {
+    if (sessionStore instanceof MeteredSessionStore) {
+      final StateSerdes<Struct, GenericRow> serdes;
+      try {
+        serdes = (StateSerdes<Struct, GenericRow>) SERDES_FIELD.get(sessionStore);
+      } catch (final IllegalAccessException e) {
+        throw new RuntimeException("Stream internals changed unexpectedly!", e);
+      }
+
+      final Bytes rawKey = Bytes.wrap(serdes.rawKey(key));
+      SessionStore<Bytes, byte[]> wrapped
+          = ((MeteredSessionStore<Struct, GenericRow>) sessionStore).wrapped();
+      while (wrapped instanceof WrappedStateStore) {
+        final StateStore store = ((WrappedStateStore<?, ?, ?>) wrapped).wrapped();
+        if (!(store instanceof SessionStore)) {
+          break;
+        }
```

Comment:
This is interesting; why does it happen?

...Ah, I see. It's because the RocksDBSessionStore wraps a SegmentedBytesStore.

The rest seems fairly obvious, but this one might be subtle enough to warrant  a comment.

Reply:
Yes, it's exactly as you mentioned.  Stopping at the last `SessionStore` is sufficient since it is below the cache and still has the expected interface for fetching.  Added a few comments about that.